include/oneapi/dpl/internal/function.h

```diff
@@ -90,6 +91,72 @@ get_access(const Policy& policy, T* ptr)
     assert(sycl::get_pointer_type(ptr, policy.queue().get_context()) == sycl::usm::alloc::shared);
     return ptr;
 }
+
+template <typename ValueType, typename Policy, typename Iterator>
+ValueType
+get_data_at(Policy p, Iterator i, size_t index)
+{
+    return get_access<sycl::access::mode::read>(p, i)[index];
+}
+
+template <typename ValueType, typename Policy, typename T>
+ValueType
+get_data_at(const Policy& policy, T* ptr, size_t index)
+{
+    sycl::queue q = policy.queue();
+
+    switch (sycl::get_pointer_type(ptr, q.get_context()))
```

Comment:
Jumping in from a user's perspective and avid Thrust user, since I ran into issues related to the missing distinction between USM and device memory: Wouldn't it make sense to make this distinction statically, i.e. provide another tag-like wrapper that behaves like a buffer iterator but takes a device pointer, to signal that it needs a copy to the host, instead of dynamically querying the SYCL runtime?
An example implementation would be [thrust::device_ptr](https://thrust.github.io/doc/classthrust_1_1device__ptr.html).